src/main/java/com/google/sps/servlets/IPGeolocationServlet.java

```diff
@@ -0,0 +1,59 @@
+package com.google.sps.servlets;
+
+import com.google.sps.dao.AnalysisDao;
+import com.google.sps.dao.GeolocationDao;
+import com.google.sps.models.PCAPAnalysis;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/ipgeolocation")
+public class IPGeolocationServlet extends HttpServlet {
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      String id = getParameter(request, "PCAPId", "");
+      PCAPAnalysis analysis = AnalysisDao.getAnalysis(id);
```

Comment:
You need to instantiate a DAO object in your init() method. That DAO instance can be a mock or can be talking to your data store or something else, based on the servlet config. Then in your servlet method here you will always call analysisDaoInstance.getAnalysis(id) irrespective of mocks or real. That will make it easier to inject mocks.

Reply:
In new commits